File: MusicSalesApp/Components/Pages/Creator/SubmitTaxForm.razor.cs
#nullable enable
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using MusicSalesApp.Common.Helpers;
using MusicSalesApp.Components.Base;
using MusicSalesApp.Helpers;
using MusicSalesApp.Models;
using System.Net.Http.Json;

namespace MusicSalesApp.Components.Pages.Creator;

public partial class SubmitTaxFormModel : BlazorBase
{
    protected bool _loading = true;
    protected bool _isCooldownActive = false;
    protected string _errorMessage = string.Empty;
    private bool _hasLoadedData = false;
    private IJSObjectReference? _jsModule;
    protected bool _isMaintenanceActive = false;
    protected DateTime? _maintenanceStartUtc = null;
    protected DateTime? _maintenanceEndUtc = null;
    protected string _maintenanceStartLocal = string.Empty;
    protected string _maintenanceEndLocal = string.Empty;
    protected string _maintenanceTimeZoneAbbreviation = string.Empty;
    private ApplicationUser? _currentUser = null;
    private bool _taxFormLoadedTracked = false;
    private bool _taxFormReturnedTracked = false;

    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender && !_hasLoadedData)
        {
            _hasLoadedData = true;
            try
            {
                var authState = await AuthenticationStateProvider.GetAuthenticationStateAsync();
                var user = authState.User;

                if (user.Identity?.IsAuthenticated != true)
                {
                    NavigationManager.NavigateTo(AppPageRoutes.Login);
                    return;
                }

                _currentUser = await UserManager.GetUserAsync(user);
                if (_currentUser == null)
                {
                    NavigationManager.NavigateTo(AppPageRoutes.Login);
                    return;
                }

                // Check if Tax Bandits is currently in a maintenance window
                if (await AppSettingsService.IsTaxBanditsMaintenanceActiveAsync())
                {
                    _isMaintenanceActive = true;
                    _maintenanceStartUtc = await AppSettingsService.GetTaxBanditsMaintenanceStartUtcAsync();
                    _maintenanceEndUtc = await AppSettingsService.GetTaxBanditsMaintenanceEndUtcAsync();

                    // Get user's local time zone info via JS interop
                    var localInfo = await TimeZoneHelper.GetUserLocalTimeAsync(JS, _maintenanceStartUtc, _maintenanceEndUtc);
                    _maintenanceStartLocal = localInfo.StartLocal;
                    _maintenanceEndLocal = localInfo.EndLocal;
                    _maintenanceTimeZoneAbbreviation = localInfo.TimeZoneAbbreviation;

                    _loading = false;
                    await InvokeAsync(StateHasChanged);
                    return;
                }

                // Check if user is in a 24-hour TIN match cooldown period
                var userIdInt = GetUserId(user);
                if (userIdInt.HasValue)
                {
                    var creator = await CreatorService.GetCreatorByUserIdAsync(userIdInt.Value);
                    if (creator?.LastTinMatchFailedAt != null)
                    {
                        var cooldownEnd = creator.LastTinMatchFailedAt.Value.AddHours(24);
                        if (DateTime.UtcNow < cooldownEnd)
                        {
                            _isCooldownActive = true;
                            _loading = false;
                            await InvokeAsync(StateHasChanged);
                            return;
                        }
                    }
                }

                // Get the transient token and configuration from the server
                Logger.LogInformation("Fetching tax form token from API");
                var response = await Http.GetFromJsonAsync<TaxFormTokenResponse>("api/creator/tax-form-token");

                if (response == null || !response.Success)
                {
                    Logger.LogWarning("Tax form token request failed: {Error}", response?.ErrorMessage ?? "null response");
                    _errorMessage = response?.ErrorMessage ?? "Failed to load tax form. Please return to Creator / Artist Settings and try again.";
                    _loading = false;
                    await InvokeAsync(StateHasChanged);
                    return;
                }

                Logger.LogInformation("Tax form token received successfully. BusinessId: {BusinessId}",
                    response.BusinessId);

                await TrackTaxFormLoadedAsync();

                _loading = false;
                await InvokeAsync(StateHasChanged);

                // Load JS module and initialize the Drop-in UI
                _jsModule = await JS.InvokeAsync<IJSObjectReference>("import", "./Components/Pages/Creator/SubmitTaxForm.razor.js");
                
                var baseUrl = NavigationManager.BaseUri.TrimEnd('/');
                Logger.LogInformation("Initializing TaxBandits Drop-in UI with return URL: {ReturnUrl}", $"{baseUrl}/CreatorSettings");
                await _jsModule.InvokeVoidAsync("initTaxForm",
                    response.TransientToken,
                    response.PayeeRef,
                    response.BusinessId,
                    response.ScriptUrl,
                    $"{baseUrl}/CreatorSettings",
                    DotNetObjectReference.Create(this));
            }
            catch (Exception ex) when (CircuitTeardown.IsExpected(ex))
            {
                // The visitor left, or the circuit dropped, while this was still awaiting.
                // Nothing is wrong and there is nobody to tell, so it must not reach the
                // Error sink - that is what emailed the admin five times on 2026-09-02.
                Logger.LogDebug(ex, "Error loading tax form");
            }
            catch (Exception ex)
            {
                Logger.LogError(ex, "Error loading tax form");
                _errorMessage = "An error occurred while loading the tax form. Please return to Creator / Artist Settings and try again.";
                _loading = false;
                await InvokeAsync(StateHasChanged);
            }
        }
    }

    /// <summary>
    /// Called from JavaScript when the tax form submission completes or is cancelled.
    /// </summary>
    [JSInvokable]
    public async Task OnTaxFormComplete(string status)
    {
        Logger.LogInformation("Tax form completed with status: {Status}", status);
        await TrackTaxFormReturnedAsync(status);
        NavigationManager.NavigateTo(AppPageRoutes.CreatorSettings, forceLoad: true);
    }

    private async Task TrackTaxFormLoadedAsync()
    {
        if (_taxFormLoadedTracked || _currentUser == null)
        {
            return;
        }

        _taxFormLoadedTracked = true;

        await TrackTaxFormFunnelEventAsync(
            FunnelAnalyticsEvents.CreatorTaxFormLoaded,
            FunnelAnalyticsLabels.TaxFormLoaded,
            null);

        await RecordTaxFormHistoryAsync(
            UserHistoryEventTypes.CreatorTaxFormLoaded,
            "Creator tax form loaded through TaxBandits.");
    }

    private async Task TrackTaxFormReturnedAsync(string status)
    {
        if (_taxFormReturnedTracked || _currentUser == null)
        {
            return;
        }

        _taxFormReturnedTracked = true;

        await TrackTaxFormFunnelEventAsync(
            FunnelAnalyticsEvents.CreatorTaxFormCompletedOrReturned,
            FunnelAnalyticsLabels.TaxFormReturned,
            status);

        await RecordTaxFormHistoryAsync(
            UserHistoryEventTypes.CreatorTaxFormCompletedOrReturned,
            $"Creator tax form completed or returned with status: {status}.");
    }

    private async Task TrackTaxFormFunnelEventAsync(string eventName, string label, string? taxFormStatus)
    {
        var payload = new Dictionary<string, object>
        {
            [FunnelAnalyticsParameters.Category] = FunnelAnalyticsLabels.CreatorCategory,
            [FunnelAnalyticsParameters.Label] = label,
            [FunnelAnalyticsParameters.Source] = FunnelAnalyticsLabels.TaxBandits
        };

        if (!string.IsNullOrWhiteSpace(taxFormStatus))
        {
            payload[FunnelAnalyticsParameters.TaxFormStatus] = taxFormStatus;
        }

        try
        {
            await JS.InvokeVoidAsync(
                GoogleAdsTrackingConfigKeys.TrackFunnelEventFunctionName,
                eventName,
                payload);
        }
        catch (Exception ex)
        {
            Logger.LogWarning(ex, "Failed to send tax form funnel analytics event {EventName}", eventName);
        }
    }

    private async Task RecordTaxFormHistoryAsync(string eventType, string description)
    {
        try
        {
            await AdminNotificationService.RecordUserHistoryAsync(
                _currentUser!.Id,
                _currentUser.Email ?? string.Empty,
                eventType,
                description);
        }
        catch (Exception ex)
        {
            Logger.LogWarning(ex, "Failed to record tax form funnel history event {EventType}", eventType);
        }
    }
}

public class TaxFormTokenResponse
{
    public bool Success { get; set; }
    public string? TransientToken { get; set; }
    public string? PayeeRef { get; set; }
    public string? BusinessId { get; set; }
    public string? ScriptUrl { get; set; }
    public string? ErrorMessage { get; set; }
}

File: MusicSalesApp/Components/Pages/Creator/SubmitTaxForm.razor.js
/**
 * TaxBandits Drop-in UI integration for embedded W-9/W-8 tax form.
 * Loads the TaxBandits dropinWhCertificate.js script dynamically and initializes the form.
 * See: https://developer.taxbandits.com/docs/WhCertificate/Authentication_Setup
 */

let dotNetRef = null;

/**
 * Dynamically loads the TaxBandits Drop-in script.
 * @param {string} scriptUrl - The URL of the TaxBandits Drop-in script.
 * @returns {Promise} Resolves when the script is loaded.
 */
function loadTaxBanditsScript(scriptUrl) {
    return new Promise((resolve, reject) => {
        // Check if loadFormWH is already available from a previous load
        if (typeof loadFormWH === 'function') {
            console.log('[TaxForm] TaxBandits script already loaded');
            resolve();
            return;
        }
        console.log('[TaxForm] Loading TaxBandits script from:', scriptUrl);
        const script = document.createElement('script');
        script.src = scriptUrl;
        script.onload = () => {
            console.log('[TaxForm] TaxBandits script loaded successfully');
            resolve();
        };
        script.onerror = () => {
            console.error('[TaxForm] Failed to load TaxBandits script from:', scriptUrl);
            reject(new Error('Failed to load TaxBandits script'));
        };
        document.head.appendChild(script);
    });
}

/**
 * Initializes the TaxBandits Drop-in UI tax form using loadFormWH().
 * @param {string} transientToken - The transient token from the server.
 * @param {string} payeeRef - The PayeeRef (email) for the recipient.
 * @param {string} businessId - The TaxBandits BusinessId.
 * @param {string} scriptUrl - The URL of the TaxBandits Drop-in script.
 * @param {string} returnUrl - The URL to redirect to after form completion.
 * @param {object} dotNetObjRef - .NET object reference for callbacks.
 */
export async function initTaxForm(transientToken, payeeRef, businessId, scriptUrl, returnUrl, dotNetObjRef) {
    dotNetRef = dotNetObjRef;

    console.log('[TaxForm] Initializing tax form with:',
        'businessId:', businessId,
        'scriptUrl:', scriptUrl,
        'returnUrl:', returnUrl,
        'tokenLength:', transientToken ? transientToken.length : 0);

    try {
        await loadTaxBanditsScript(scriptUrl);

        // Listen for messages from the TaxBandits iframe
        window.addEventListener('message', handleTaxBanditsMessage);

        // Build the payload per TaxBandits API documentation
        const payLoad = {
            Requester: {
                BusinessId: businessId
            },
            Recipient: {
                PayeeRef: payeeRef,
                IsTINMatching: false
            },
            RedirectUrls: {
                ReturnUrl: returnUrl,
                CancelUrl: returnUrl
            }
        };

        console.log('[TaxForm] Calling loadFormWH with payload keys:', Object.keys(payLoad));

        // loadFormWH(transientToken, payLoad) is the API from dropinWhCertificate.js
        // Don't await — loadFormWH renders an iframe and may not resolve until the user
        // completes/cancels the form, which would cause Blazor's JS interop to time out.
        loadFormWH(transientToken, payLoad).then(() => {
            console.log('[TaxForm] loadFormWH completed successfully');
        }).catch((err) => {
            console.error('[TaxForm] loadFormWH error:', err);
        });
    } catch (error) {
        console.error('[TaxForm] Error initializing TaxBandits form:', error);
    }
}

/**
 * Handles messages from the TaxBandits iframe.
 */
function handleTaxBanditsMessage(event) {
    // TaxBandits sends messages from the iframe about form status
    if (event.data && dotNetRef) {
        try {
            const data = typeof event.data === 'string' ? JSON.parse(event.data) : event.data;
            console.log('[TaxForm] Received message from iframe:', JSON.stringify(data));
            if (data.status === 'completed' || data.status === 'cancelled') {
                dotNetRef.invokeMethodAsync('OnTaxFormComplete', data.status);
            }
        } catch {
            // Not a JSON message or not from TaxBandits, ignore
        }
    }
}
